include/alpaka/atomic/AtomicAtomicRef.hpp

```diff
@@ -0,0 +1,231 @@
+/* Copyright 2021 Felice Pantaleo (CERN), Andrea Bocci (CERN)
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/atomic/Traits.hpp>
+#include <alpaka/core/BoostPredef.hpp>
+
+#include <boost/atomic.hpp>
+
+#include <array>
+#include <type_traits>
+
+namespace alpaka
+{
+    //! The CPU threads accelerator atomic ops based on atomic_ref.
+    //
+    //  Atomics can be used in the grids, blocks and threads hierarchy levels.
+    //
+
+    class AtomicAtomicRef
+    {
+    };
+
+    namespace traits
+    {
+        //! The CPU threads accelerator AtomicAdd.
+        template<typename T, typename THierarchy>
+        struct AtomicOp<AtomicAdd, AtomicAtomicRef, T, THierarchy>
+        {
+            ALPAKA_FN_HOST static auto atomicOp(AtomicAtomicRef const& atomic, T* const addr, T const& value) -> T
+            {
+                static_assert(
+                    std::is_trivially_copyable_v<T> && boost::atomic_ref<T>::required_alignment <= alignof(T),
+                    "Type not supported by AtomicAtomicRef, please use AtomicStdLibLock for your accelerator.");
```

Comment:
Are we able to use other atomics than specified in the atomic hierarchy? Naively I would assume you would need to switch to another accelerator entirely.

Reply:
Not that we could find: the hierarchy is defined by the accelerator, and is not templated on the data type.
The comment is supposed to suggest that the user has run into a corner case not supported by `boost::atomic_ref`, and should indeed modify the accelerator to use `AtomicStdLibLock` instead of `AtomicAtomicRef`.

I think it could be done adding something like
```c++
        public AtomicHierarchy<
#ifndef ALPAKA_DISABLE_ATOMIC_ATOMICREF
            AtomicAtomicRef, // grid atomics
            AtomicAtomicRef, // block atomics
#else
            AtomicStdLibLock<16>, // grid atomics
            AtomicStdLibLock<16>, // block atomics
#endif  // ALPAKA_DISABLE_ATOMIC_ATOMICREF
            AtomicNoOp         // thread atomics
        >,
```
to all the affected accelerators, and change the comment here to suggest setting `-DALPAKA_DISABLE_ATOMIC_ATOMICREF` to fall back to `AtomicStdLibLock`.

Or, since that involves repeating the same `#ifdef` in many files and in multiple places per file, we could add it to a single place in a new file:
`include/alpaka/atomic/AtomicCpu.hpp`:
```c++
#pragma once

#ifndef ALPAKA_DISABLE_ATOMIC_ATOMICREF
#include <alpaka/atomic/AtomicAtomicRef.hpp>
#else
#include <alpaka/atomic/AtomicStdLibLock.hpp>
#endif  // ALPAKA_DISABLE_ATOMIC_ATOMICREF

namespace alpaka {

#ifndef ALPAKA_DISABLE_ATOMIC_ATOMICREF
  using AtomicCpu = AtomicAtomicRef;
#else
  using AtomicCpu = AtomicStdLibLock<16>;
#endif  // ALPAKA_DISABLE_ATOMIC_ATOMICREF

}  // namespace alpaka
```

What do you think ?

Reply:
I agree with this latter solution if someone really needs fallbacks. This might be needed just for the atomicCAS of for some embedded software applications that deal with packed/non-aligned data structures. 
If you agree, I will integrate this proposed solution in the PR.

Reply:
I don't see problem with the latter approach. Feel free to give it a try :-)

Reply:
I am not sure if `#ifdef` is the right way. Control how an accelerator behaves should be managed via policies but to implement it somehow clean IMO [accelerator tags](https://github.com/alpaka-group/alpaka/issues/1246) are required.

I do not see that we will get a non preprocessor way ready soon therefore I agree with the second solution :-(